### PR TITLE
[InstCombine] Infer nusw + nneg -> nuw for getelementptr

### PR DESCRIPTION
**Pull request by nikic**
If the gep is nusw (usually via inbounds) and the offset is non-negative, we can infer nuw.

Unfortunately this inference does have some compile-time overhead: https://llvm-compile-time-tracker.com/compare.php?from=37e5319a12ba47c18049728804d3d1e1b10c4eb4&to=af56d73d6543f05b1e5205b96934e2427bb24d72&stat=instructions:u

Proof: https://alive2.llvm.org/ce/z/ihztLy

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 1]
```diff
@@ -3096,6 +3096,15 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
     }
   }
 
+  // nusw + nneg -> nuw
+  if (GEP.hasNoUnsignedSignedWrap() && !GEP.hasNoUnsignedWrap() &&
+      all_of(GEP.indices(), [&](Value *Idx) {
+        return isKnownNonNegative(Idx, SQ.getWithInstruction(&GEP));
+      })) {
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:3103]: Any alleviation of compile time impact if you do only constants first to rule out trivial case?

**Review by goldsteinn — APPROVED** [review]
LGTM although might be a bit more we do to reduce compile time impact.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 2]
```diff
@@ -3096,6 +3096,15 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
...
+    GEP.setNoWrapFlags(GEP.getNoWrapFlags() | GEPNoWrapFlags::noUnsignedWrap());
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:3104]: Should drop duplicate logic in other places:
```
SelectionDAGBuilder::visitGetElementPtr
InstCombinerImpl::visitPtrToInt
llvm/lib/Transforms/Scalar/LICM.cpp:hoistGEP
SeparateConstOffsetFromGEP::reorderGEP
```

**Comment by nikic** [comment]
> LGTM although might be a bit more we do to reduce compile time impact.

I don't think there is much than we can do here apart from a generic "make computeKnownBits faster" :(

**Pull request merged**